File: data-stack/to_be_deleted/ibc2025-mcp-data-generation-repo/cdk/stacks/data_lake_stack.py
import aws_cdk as cdk
from aws_cdk import (
    Stack,
    aws_s3 as s3,
    aws_iam as iam,
    RemovalPolicy,
    Duration,
    CfnOutput
)
from constructs import Construct

class DataLakeStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)
        
        # Create S3 bucket for data lake
        self.data_bucket = s3.Bucket(
            self,
            "DataLakeBucket",
            bucket_name=f"acme-streaming-data-lake-{self.account}-{self.region}",
            versioned=True,
            encryption=s3.BucketEncryption.S3_MANAGED,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            removal_policy=RemovalPolicy.DESTROY,  # Change to RETAIN for production
            auto_delete_objects=True,  # Remove for production
            lifecycle_rules=[
                s3.LifecycleRule(
                    id="TransitionToIA",
                    transitions=[
                        s3.Transition(
                            storage_class=s3.StorageClass.INFREQUENT_ACCESS,
                            transition_after=Duration.days(30)
                        ),
                        s3.Transition(
                            storage_class=s3.StorageClass.GLACIER,
                            transition_after=Duration.days(90)
                        )
                    ]
                ),
                s3.LifecycleRule(
                    id="DeleteOldVersions",
                    noncurrent_version_expiration=Duration.days(30),
                    abort_incomplete_multipart_upload_after=Duration.days(7)
                )
            ],
            cors=[
                s3.CorsRule(
                    allowed_methods=[
                        s3.HttpMethods.GET,
                        s3.HttpMethods.HEAD,
                        s3.HttpMethods.PUT
                    ],
                    allowed_origins=["*"],
                    allowed_headers=["*"],
                    max_age=3000
                )
            ]
        )
        
        # Create folder structure in S3
        self._create_folder_structure()
        
        # Create IAM role for data access
        self.data_access_role = iam.Role(
            self,
            "DataAccessRole",
            assumed_by=iam.CompositePrincipal(
                iam.ServicePrincipal("glue.amazonaws.com"),
                iam.ServicePrincipal("athena.amazonaws.com"),
                iam.ServicePrincipal("redshift.amazonaws.com")
            ),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AWSGlueServiceRole")
            ]
        )
        
        # Grant read/write permissions to the role
        self.data_bucket.grant_read_write(self.data_access_role)
        
        # Create a bucket policy for cross-service access
        bucket_policy = s3.BucketPolicy(
            self,
            "DataLakeBucketPolicy",
            bucket=self.data_bucket
        )
        
        bucket_policy.document.add_statements(
            iam.PolicyStatement(
                effect=iam.Effect.ALLOW,
                principals=[
                    iam.ServicePrincipal("athena.amazonaws.com"),
                    iam.ServicePrincipal("glue.amazonaws.com"),
                    iam.ServicePrincipal("redshift.amazonaws.com")
                ],
                actions=[
                    "s3:GetObject",
                    "s3:ListBucket",
                    "s3:GetBucketLocation"
                ],
                resources=[
                    self.data_bucket.bucket_arn,
                    f"{self.data_bucket.bucket_arn}/*"
                ]
            )
        )
        
        # Outputs
        CfnOutput(
            self,
            "DataBucketName",
            value=self.data_bucket.bucket_name,
            description="Name of the S3 data lake bucket"
        )
        
        CfnOutput(
            self,
            "DataBucketArn",
            value=self.data_bucket.bucket_arn,
            description="ARN of the S3 data lake bucket"
        )
        
        CfnOutput(
            self,
            "DataAccessRoleArn",
            value=self.data_access_role.role_arn,
            description="ARN of the data access IAM role"
        )
    
    def _create_folder_structure(self):
        # Note: S3 doesn't have real folders, but we can create zero-byte objects
        # with trailing slashes to simulate folder structure in the console
        folders = [
            "raw/customers/",
            "raw/titles/",
            "raw/telemetry/",
            "raw/campaigns/",
            "processed/customers/",
            "processed/titles/",
            "processed/telemetry/",
            "processed/campaigns/",
            "analytics/reports/",
            "analytics/dashboards/",
            "temp/"
        ]
        
        # In practice, these folders will be created when data is uploaded
        # This is just for documentation purposes

File: data-stack/consolidated-data-stack/lib/stacks/data-lake-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as glue from 'aws-cdk-lib/aws-glue';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';
import { Config } from '../config';

export class DataLakeStack extends cdk.Stack {
  public readonly dataBucket: s3.Bucket;
  public readonly glueDatabase: glue.CfnDatabase;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // S3 bucket for telemetry data
    this.dataBucket = new s3.Bucket(this, 'DataBucket', {
      bucketName: `${Config.s3.dataBucketName}-${this.account}-${this.region}`,
      encryption: s3.BucketEncryption.S3_MANAGED,
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      lifecycleRules: [
        {
          id: 'MoveToIA',
          transitions: [
            {
              storageClass: s3.StorageClass.INFREQUENT_ACCESS,
              transitionAfter: cdk.Duration.days(30),
            },
            {
              storageClass: s3.StorageClass.GLACIER,
              transitionAfter: cdk.Duration.days(90),
            },
          ],
        },
        {
          id: 'DeleteOldVersions',
          noncurrentVersionExpiration: cdk.Duration.days(30),
        },
        {
          id: 'AbortIncompleteUploads',
          abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
        },
      ],
    });

    // Glue database
    this.glueDatabase = new glue.CfnDatabase(this, 'GlueDatabase', {
      catalogId: this.account,
      databaseInput: {
        name: Config.glue.databaseName,
        description: 'ACME video streaming telemetry data',
      },
    });

    // Glue table for streaming events
    const glueTable = new glue.CfnTable(this, 'StreamingEventsTable', {
      catalogId: this.account,
      databaseName: Config.glue.databaseName,
      tableInput: {
        name: Config.glue.tableName,
        description: 'Video streaming telemetry events',
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          'classification': 'parquet',
          'parquet.compression': 'SNAPPY',
        },
        storageDescriptor: {
          location: `s3://${this.dataBucket.bucketName}/telemetry/`,
          inputFormat: 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat',
          outputFormat: 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat',
          serdeInfo: {
            serializationLibrary: 'org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe',
          },
          columns: [
            { name: 'event_id', type: 'string' },
            { name: 'event_type', type: 'string' },
            { name: 'event_timestamp', type: 'string' },
            { name: 'customer_id', type: 'string' },
            { name: 'title_id', type: 'string' },
            { name: 'session_id', type: 'string' },
            { name: 'device_id', type: 'string' },
            { name: 'title_type', type: 'string' },
            { name: 'device_type', type: 'string' },
            { name: 'device_os', type: 'string' },
            { name: 'app_version', type: 'string' },
            { name: 'quality', type: 'string' },
            { name: 'bandwidth_mbps', type: 'double' },
            { name: 'buffering_events', type: 'int' },
            { name: 'buffering_duration_seconds', type: 'double' },
            { name: 'error_count', type: 'int' },
            { name: 'watch_duration_seconds', type: 'int' },
            { name: 'position_seconds', type: 'int' },
            { name: 'completion_percentage', type: 'double' },
            { name: 'ip_address', type: 'string' },
            { name: 'isp', type: 'string' },
            { name: 'connection_type', type: 'string' },
            { name: 'country', type: 'string' },
            { name: 'state', type: 'string' },
            { name: 'city', type: 'string' },
          ],
        },
        partitionKeys: [
          { name: 'year', type: 'string' },
          { name: 'month', type: 'string' },
          { name: 'day', type: 'string' },
          { name: 'hour', type: 'string' },
        ],
      },
    });
    glueTable.addDependency(this.glueDatabase);

    // IAM role for Athena/Glue access
    const dataAccessRole = new iam.Role(this, 'DataAccessRole', {
      roleName: `${Config.prefix}-data-access-role`,
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal('glue.amazonaws.com'),
        new iam.ServicePrincipal('athena.amazonaws.com')
      ),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSGlueServiceRole'),
      ],
    });

    this.dataBucket.grantReadWrite(dataAccessRole);

    // Outputs
    new cdk.CfnOutput(this, 'DataBucketName', { value: this.dataBucket.bucketName });
    new cdk.CfnOutput(this, 'GlueDatabaseName', { value: Config.glue.databaseName });
  }
}
